debian/man/zimrecreate.1
.TH ZIMRECREATE "1" "July 2020" "zim-tools" "User Commands"
.SH NAME
zimrecreate \- tool to recreate a ZIM files from a existing ZIM
.SH SYNOPSIS
\fBzimpatch\fR [options] origin_file output_file\fR
.SH DESCRIPTION
.TP
origin_file
Original file
.TP
output_file
File to output to
.TP
\fB\-h\fR, \fB\-\-help\fR
show a help message and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
print software version



ChangeLog
zim-tools 2.0.0
===============

 * zimwriterfs: the tool version 1.3.10 migrated to zim-tools sources tree
 * zimwriterfs: Better support of webp format
 * zimdump: Fix directory and redirection handling when dumping zim content of
   the file system.
 * zimdump: Improve zimdump commandline argument.
 * zimcheck: Fix false positive with URL containing '?'
 * Fix help typo in zimpatch and zimcheck
 * Fix incoherent licensing.

zim-tools 1.2.1
===============

 * Fix compilation warning

zim-tools 1.2
=============

 * Fix a lot small issues about CI compilation.
 * zimdump: Add an option to dump redirection using html redirect or symlink.

zim-tools 1.1.1
===============

 * Adapt to new api of libzim.

1.1.0
=====

 * [CI] Move to github's action ci.
 * zimdump: Use the url article instead of the title as filename.
 * zimbench: Fix time measurement.

1.0.5
=====

 * Adapt to the new libzim's creator api
 * zimrecreate: Better performance.
 * zimcheck: Check if an entry is empty

1.0.4
=====

 * zimcheck: Better handling of ./ in URL checking

1.0.3
=====

 * ZimCheck handles properly geo: mailto: tel: javascript: links
 * Fix external URL checking error message and usage()
 * Added -V --version command line arguments

1.0.2
=====

 * Update readme.
 * ZimCheck returns a proper status code.
 * Update to last API of libzim 5.0.0
 * Temporally deactivate compilation of zimBench on windows.

1.0.1
=====

 * Update build.
 * [zimCheck] Far better zimCheck implementation :
    - performance issue
    - fix url check
 * Build without rpath

1.0.0
=====

 * Remove warnings.
 * Compile zim-tools with travis.
 * Add zim-check tool.
 * Addapt zimDiff and zimPatch to new version of libzim writer.
 * [zimSplit] Do not create a new ofstream for each part. (Do not need gcc5)
 * Do not use `std::to_string` (Allow compilation with gcc4.8)


src/zimrecreate.cpp
/*
 * Copyright (C) 2013 Kiran Mathew Koshy
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU  General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 */

#include <iostream>
#include <sstream>
#include <vector>
#include <zim/writer/creator.h>
#include <zim/blob.h>
#include <zim/item.h>
#include <zim/archive.h>
#include <list>
#include <algorithm>
#include <sstream>

#include "version.h"

class Article : public zim::writer::Article         //Article class that will be passed to the zimwriter. Contains a zim::Article class, so it is easier to add a
{
    zim::Entry entry;
public:
    explicit Article(const zim::Entry e):
      entry(e)
    {}

    virtual zim::writer::Url getUrl() const
    {
        return zim::writer::Url(entry.getPath());
    }

    virtual std::string getTitle() const
    {
        return entry.getTitle();
    }

    virtual bool isRedirect() const
    {
        return entry.isRedirect();
    }

    virtual std::string getMimeType() const
    {
        if (isRedirect()) { return ""; }
        return entry.getItem().getMimetype();
    }

    virtual zim::writer::Url getRedirectUrl() const {
      auto redirectEntry = entry.getRedirectEntry();
      return zim::writer::Url(redirectEntry.getPath());
    }

    zim::Blob getData() const
    {
        return entry.getItem().getData();
    }

    zim::size_type getSize() const
    {
        return entry.getItem().getSize();
    }

    std::string getFilename() const
    {
        return "";
    }

    bool shouldCompress() const
    {
        return getMimeType().find("text") == 0
            || getMimeType() == "application/javascript"
            || getMimeType() == "application/json"
            || getMimeType() == "image/svg+xml";
    }

    bool shouldIndex() const
    {
        return getMimeType().find("text/html") == 0;
    }
};

using pair_type = std::pair<zim::entry_index_type, zim::cluster_index_type>;

class ComparatorByCluster {
  public:
    ComparatorByCluster(const zim::Archive& origin):
      origin(origin) {
    }

    bool operator() (pair_type i, pair_type j) {
      return i.second < j.second;
    }
  const zim::Archive& origin;
};


class ZimRecreator : public zim::writer::Creator
{
    zim::Archive origin;

public:
    explicit ZimRecreator(std::string originFilename="") :
      zim::writer::Creator(true),
      origin(originFilename)
    {
        // [TODO] Use the correct language
        setIndexing(true, "eng");
        setMinChunkSize(2048);
    }

    virtual void create(const std::string& fname)
    {
        std::cout << "starting zim creation" << std::endl;
        startZimCreation(fname);
        for(auto& entry:origin.iterEfficient())
        {
          auto path = entry.getPath();
          if (path[0] == 'Z' || path[0] == 'X') {
            // Index is recreated by zimCreator. Do not add it
            continue;
          }
          auto tempArticle = std::make_shared<Article>(entry);
          addArticle(tempArticle);
        }
        finishZimCreation();
    }

    virtual zim::writer::Url getMainUrl() const {
      try {
        return zim::writer::Url(origin.getMainEntry().getPath());
      } catch(...) {
        return zim::writer::Url();
      }
    }

    virtual zim::writer::Url getLayoutUrl() const {
      return zim::writer::Url();
    }
};

void displayHelp()
{
    std::cout<<"\nzimrecreate"
    "\nA tool to recreate a ZIM files from a existing ZIM."
    "\nUsage: zimrecreate [origin_file] [output file]"
    "\nOption: -v, --version    print software version\n";
    return;
}

int main(int argc, char* argv[])
{

    //Parsing arguments
    //There will be only two arguments, so no detailed parsing is required.
    std::cout<<"zimrecreate\n";
    for(int i=0;i<argc;i++)
    {
        if(std::string(argv[i])=="-H" ||
           std::string(argv[i])=="--help" ||
           std::string(argv[i])=="-h")
        {
            displayHelp();
            return 0;
        }

        if(std::string(argv[i])=="--version" ||
           std::string(argv[i])=="-v")
        {
            version();
            return 0;
        }
    }
    if(argc<3)
    {
        std::cout<<"\n[ERROR] Not enough Arguments provided\n";
        return -1;
    }
    std::string originFilename =argv[1];
    std::string outputFilename =argv[2];
    try
    {
        ZimRecreator c(originFilename);
        //Create the actual file.
        c.create(outputFilename);
    }
    catch (const std::exception& e)
    {
        std::cerr << e.what() << std::endl;
    }
}
